engine/math/vector.lisp

```diff
@@ -0,0 +1,249 @@
+
+(defmethod sub ((vector1 float4) vector2)
+  (make-float4 (- (x vector1) (x vector2))
+	       (- (y vector1) (y vector2))
+	       (- (z vector1) (z vector2))
+	       (- (w vector1) (w vector2))))
+
+
+(defmethod mul ((vector float2) scalar)
+  (make-float2 (* (x vector) scalar)
+	       (* (y vector) scalar)))
+
+(defmethod mul ((vector float3) scalar)
+  (make-float3 (* (x vector) scalar)
+	       (* (y vector) scalar)
+	       (* (z vector) scalar)))
+
+(defmethod mul ((vector float4) scalar)
+  (make-float4 (* (x vector) scalar)
+	       (* (y vector) scalar)
+	       (* (z vector) scalar)
+	       (* (w vector) scalar)))
+
+
+(defmethod div ((vector float2) scalar)
+  (make-float2 (/ (x vector) scalar)
+	       (/ (y vector) scalar)))
+
+(defmethod div ((vector float3) scalar)
+  (make-float3 (/ (x vector) scalar)
+	       (/ (y vector) scalar)
+	       (/ (z vector) scalar)))
+
+(defmethod div ((vector float4) scalar)
+  (make-float4 (/ (x vector) scalar)
+	       (/ (y vector) scalar)
+	       (/ (z vector) scalar)
+	       (/ (w vector) scalar)))
+
+
+(defmethod dot ((vector1 float2) vector2)
+  (+ (* (x vector1) (x vector2))
+     (* (y vector1) (y vector2))))
+
+(defmethod dot ((vector1 float3) vector2)
+  (+ (* (x vector1) (x vector2))
+     (* (y vector1) (y vector2))
+     (* (z vector1) (z vector2))))
+
+(defmethod dot ((vector1 float4) vector2)
+  (+ (* (x vector1) (x vector2))
+     (* (y vector1) (y vector2))
+     (* (z vector1) (z vector2))
+     (* (w vector1) (w vector2))))
+
+
+(defmethod length ((vector float2))
+  (sqrt (+ (expt (x vector) 2)
+	   (expt (y vector) 2))))
+  
+(defmethod length ((vector float3))
+  (sqrt (+ (expt (x vector) 2)
+	   (expt (y vector) 2)
+	   (expt (z vector) 2))))
+  
+(defmethod length ((vector float4))
+  (sqrt (+ (expt (x vector) 2)
+	   (expt (y vector) 2)
+	   (expt (z vector) 2)
+	   (expt (w vector) 2))))
+
+
+(defmethod eq ((vector1 float2) vector2)
+  (and (= (x vector1) (x vector2))
+       (= (y vector1) (y vector2))))
+
+(defmethod eq ((vector1 float3) vector2)
+  (and (= (x vector1) (x vector2))
+       (= (y vector1) (y vector2))
+       (= (z vector1) (z vector2))))
+
+(defmethod eq ((vector1 float2) vector2)
+  (and (= (x vector1) (x vector2))
+       (= (y vector1) (y vector2))
+       (= (z vector1) (z vector2))
+       (= (w vector1) (w vector2))))
+
+
+(defun cross (vector1 vector2)
+  (make-float3 (- (* (y vector1) (z vector2))
+		  (* (z vector1) (y vector2)))
+	       (- (* (z vector1) (x vector2))
+		  (* (x vector1) (z vector2)))
+	       (- (* (x vector1) (y vector2))
+		  (* (y vector1) (x vector2))))) 
+
+
+(defmethod normalize (vector)
+  (div vector (length vector)))
+
+
+(defmethod angle (vector1 vector2)
+  (acos (/ (dot vector1 vector 2)
+	   (* (length vector1)
+	      (length vector2)))))
+
+
+(defmethod project (vector1 vector2)
+  (dot vector1 (normalize vector2)))
+
+
+(defvar +float2-zero+ (make-instance 'float2))
```

Comment:
Using `defvar` might not be the best idea, because this single instance defined in the constant will be used everywhere (since classes are passed by-reference).

Reply:
Should I `defun` a function that returns a given thing instead?